File: app/src/main/java/org/wikipedia/json/WikiSiteTypeAdapter.java
package org.wikipedia.json;

import android.net.Uri;

import com.google.gson.JsonParseException;
import com.google.gson.TypeAdapter;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonToken;
import com.google.gson.stream.JsonWriter;

import org.wikipedia.dataclient.WikiSite;

import java.io.IOException;

public class WikiSiteTypeAdapter extends TypeAdapter<WikiSite> {
    private static final String DOMAIN = "domain";
    private static final String LANGUAGE_CODE = "languageCode";

    @Override public void write(JsonWriter out, WikiSite value) throws IOException {
        out.beginObject();
        out.name(DOMAIN);
        out.value(value.url());

        out.name(LANGUAGE_CODE);
        out.value(value.languageCode());
        out.endObject();
    }

    @Override public WikiSite read(JsonReader in) throws IOException {
        // todo: legacy; remove in June 2018
        if (in.peek() == JsonToken.STRING) {
            return new WikiSite(Uri.parse(in.nextString()));
        }

        String domain = null;
        String languageCode = null;
        in.beginObject();
        while (in.hasNext()) {
            String field = in.nextName();
            String val = in.nextString();
            switch (field) {
                case DOMAIN:
                    domain = val;
                    break;
                case LANGUAGE_CODE:
                    languageCode = val;
                    break;
                default: break;
            }
        }
        in.endObject();

        if (domain == null) {
            throw new JsonParseException("Missing domain");
        }

        // todo: legacy; remove in June 2018
        if (languageCode == null) {
            return new WikiSite(domain);
        }
        return new WikiSite(domain, languageCode);
    }
}

File: app/src/main/java/org/wikipedia/json/WikiSiteTypeAdapter.kt
package org.wikipedia.json

import com.google.gson.JsonParseException
import com.google.gson.TypeAdapter
import com.google.gson.stream.JsonReader
import com.google.gson.stream.JsonWriter
import org.wikipedia.dataclient.WikiSite
import java.io.IOException

class WikiSiteTypeAdapter : TypeAdapter<WikiSite>() {

    @Throws(IOException::class)
    override fun write(writer: JsonWriter, value: WikiSite) {
        writer.beginObject()
        writer.name(DOMAIN)
        writer.value(value.url())
        writer.name(LANGUAGE_CODE)
        writer.value(value.languageCode())
        writer.endObject()
    }

    @Throws(IOException::class)
    override fun read(reader: JsonReader): WikiSite {
        var domain: String? = null
        var languageCode: String? = null
        reader.beginObject()
        while (reader.hasNext()) {
            val field = reader.nextName()
            val value = reader.nextString()
            when (field) {
                DOMAIN -> domain = value
                LANGUAGE_CODE -> languageCode = value
            }
        }
        reader.endObject()
        if (domain.isNullOrEmpty()) {
            throw JsonParseException("Missing domain")
        }

        return languageCode?.let { WikiSite(domain, it) } ?: WikiSite(domain)
    }

    companion object {
        private const val DOMAIN = "domain"
        private const val LANGUAGE_CODE = "languageCode"
    }
}
